lib/models/requestResponse.js

```diff
@@ -0,0 +1,242 @@
+'use strict';
+
+const
+  assert = require('../utils/assertType');
+
+const
+  _request = Symbol('request'),
+  _headers = Symbol('headers');
+
+/**
+ * Kuzzle normalized response
+ *
+ * @class
+ * @param {Request} request
+ *
+ */
+class RequestResponse {
+  constructor (request) {
+
+    this[_request] = request;
+    this[_headers] = this[_request][Symbol.for('request.response.headers')];
+
+    Object.seal(this);
+  }
+
+  /**
+   * Get the parent request status
+   * @returns {number}
+   */
+  get status () {
+    return this[_request].status;
+  }
+
+  /**
+   * Set the parent request status
+   * @param {number} s
+   */
+  set status (s) {
+    return this[_request].status = s;
+  }
+
+  /**
+   * Get the parent request error
+   * @returns {KuzzleError}
+   */
+  get error () {
+    return this[_request].error;
+  }
+
+  /**
+   * Set the parent request error
+   * @param {KuzzleError} e
+   */
+  set error (e) {
+    return this[_request].setError(e);
+  }
+
+  /**
+   * Get the parent request id
+   * @returns {string|*|String}
+   */
+  get requestId () {
+    return this[_request].id;
+  }
+
+  /**
+   * Get the parent request controller
+   * @returns {string}
+   */
+  get controller () {
+    return this[_request].input.controller;
+  }
+
+  /**
+   * Get the parent request action
+   * @returns {string}
+   */
+  get action () {
+    return this[_request].input.action;
+  }
+
+  /**
+   * Get the parent request collection
+   * @returns {string}
+   */
+  get collection () {
+    return this[_request].input.resource.collection;
+  }
+
+  /**
+   * Get the parent request index
+   * @returns {string}
+   */
+  get index () {
+    return this[_request].input.resource.index;
+  }
+
+  /**
+   * Get the parent request metadata
+   * @returns {Object}
+   */
+  get metadata () {
+    return this[_request].input.metadata;
+  }
+
+  /**
+   * Get the response headers - reference to private parent request property
+   * @returns {*}
+   */
+  get headers () {
+    return this[_headers];
+  }
+
+  /**
+   * Get the parent request result
+   * @returns {*|null|*|Object}
+   */
+  get result () {
+    return this[_request].result;
+  }
+
+  /**
+   * Set the parent request result
+   * @param {*} r
+   */
+  set result (r) {
+    return this[_request].setResult(r);
+  }
+
+  /**
+   * Get the header value for {name} (case-insensitive)
+   * @public
+   * @param {string} name
+   */
+  getHeader (name) {
+    let key = Object.keys(this[_headers])
+      .filter(key => key.toLowerCase() === name.toLowerCase())[0];
```

Comment:
I agree on the cases where we need to retrieve the key, either to return or set the value, not on the deletion cases.
Even if quite unlikely, technically, it is still possible to access the internal headers collection and alter it as I had to make the Symbol global. In case a duplicate is inserted, the removal and de-duplication should still work.

Updated in this way.